rust/libceed/src/qfunction.rs

```diff
@@ -191,6 +278,71 @@ impl<'a> QFunctionCore<'a> {
         };
         self.ceed.check_error(ierr)
     }
+
+    pub fn inputs(&self) -> crate::Result<Vec<crate::QFunctionField>> {
+        // Get number of inputs
+        let mut num_inputs = 0;
+        let ierr = unsafe {
+            bind_ceed::CeedQFunctionGetNumArgs(
+                self.ptr,
+                &mut num_inputs,
+                std::ptr::null_mut() as *mut i32,
+            )
+        };
+        self.ceed.check_error(ierr)?;
+        // Get array of raw C pointers for inputs
+        let mut inputs_ptr = std::ptr::null_mut();
+        let ierr = unsafe {
+            bind_ceed::CeedQFunctionGetFields(
+                self.ptr,
+                &mut inputs_ptr,
+                std::ptr::null_mut() as *mut *mut bind_ceed::CeedQFunctionField,
+            )
+        };
+        self.ceed.check_error(ierr)?;
+        // Convert raw C pointers to fixed length array
+        let inputs_slice = unsafe { std::slice::from_raw_parts(inputs_ptr, num_inputs as usize) };
+        let mut inputs = Vec::new();
+        for i in 0..num_inputs as usize {
+            inputs.push(QFunctionField {
+                ptr: inputs_slice[i] as bind_ceed::CeedQFunctionField,
+            })
+        }
+        Ok(inputs)
+    }
+
+    pub fn outputs(&self) -> crate::Result<Vec<crate::QFunctionField>> {
```

Comment:
I'm a little uncomfortable with this returning *owned* `QFunctionField` (modifications of which would be reflected as mutation on the C side). Wouldn't it be more natural to an immutable slice?

Reply:
I think if these interfaces can be turned into the following, we'll be good with this PR. Unless you have a rationale for for these being owned `Vec`.

```suggestion
    pub fn outputs(&self) -> crate::Result<&[crate::QFunctionField]> {
```

Reply:
I'm currently digging through this. The tricky thing is that the slice has to be owned *somewhere*. I'm looking at the interior mutability pattern in Rust, but it is tricky to make a clean return of a slice instead of a ref. It might be easier to return a `QFunctionFieldView` that only has immutable deref.

Reply:
Current working RefCell version
```
    pub fn inputs(&self) -> crate::Result<&[crate::QFunctionField]> {
        // Get array of raw C pointers for inputs
        let mut num_inputs = 0;
        let mut inputs_ptr = std::ptr::null_mut();
        let ierr = unsafe {
            bind_ceed::CeedQFunctionGetFields(
                self.ptr,
                &mut num_inputs,
                &mut inputs_ptr,
                std::ptr::null_mut() as *mut bind_ceed::CeedInt,
                std::ptr::null_mut() as *mut *mut bind_ceed::CeedQFunctionField,
            )
        };
        self.ceed.check_error(ierr)?;
        // Convert raw C pointers to fixed length array
        let inputs_slice = unsafe { std::slice::from_raw_parts(inputs_ptr, num_inputs as usize) };
        let mut inputs = Vec::new();
        for i in 0..num_inputs as usize {
            inputs.push(QFunctionField {
                ptr: inputs_slice[i] as bind_ceed::CeedQFunctionField,
            })
        }
        self.inputs.replace(inputs);
        Ok(unsafe { self.inputs.try_borrow_unguarded().unwrap() })
    }
```

Reply:
Since `QFunctionField` just wraps a single pointer, it seems you could just transmute it. Then you wouldn't need an owned `Vec` or to even traverse the array (trusting that the pointers are non-NULL) because everything is contained in the slice, and the backing memory for the slice lives as long as the `QFunctionCore<'a>`.

Reply:
I think the reasoning about this gets easier if I restrict `*GetFields()` to only be valid if the Operator is fully set-up.

Reply:
Good point, and I think that's a reasonable constraint. (`QFunctionGetFields` requires a complete `CeedQFunction` and similar for `CeedOperator`.)

Reply:
It is easy to check if a `CeedOperator` is fully specified - we check if it has as many fields as the `CeedQFunction`

Currently, we don't have a way to check that all `CeedQFunction` fields have been added. Should our QFunction constructor specify the number of inputs and outputs in the C interface?

Reply:
Could we make it so that doing anything non-creationary with a QFunction checks that it's complete (to the extent that's a thing) and makes it immutable? This would avoiding making the API more burdensome, but still prevent weird conditions.

Reply:
Currently we make the QF immutable once the OP is successfully checked for 'completeness' prior to Apply, Assemble*, etc. We don't really have any way of knowing that a QFunction is complete on its own.

Reply:
But if all you have is a QFunction, we'd still like to be able to ask it questions. Can we mark it immutable in the same way as would happen when an Operator checks it? (Bottom line is that we shouldn't need an Operator to put the QFunction in that state.) It could be `CeedQFunctionVerify()` that the user may optionally call and that we call for certain queries and when Operator uses it.

Reply:
But how do we know all fields have been provided unless the user tells us how many fields should be provided? Unless we just assume that if the user calls GetFields or Apply then they are no longer allowed to make any modifications.

Reply:
Right, we just document the semantic and the user will get an error if they try to add more fields later.

Reply:
Ok, I've rearranged this PR a bit. I still need to play more with returning a slice, though what I wrote above should be perfectly fine now that the fields cannot be changed after we call *GetFields

There is still some Julia issue I need to figure out, but that's for tomorrow.

Reply:
Nope, the code above won't work. Rust will let you take the immutable borrow in the return value and put it in a mutable variable and then change properties of the immutable borrow willy-nilly.

Reply:
Ah, new plan